userguide/tutorials/notification_plugin.adoc

```diff
@@ -1,31 +1,142 @@
 
-By default, if your plugin triggers a runtime exception, the event is redispatched right away up to 3 times (or as configured by the `org.killbill.notificationq.external.max.failure.retry` global property).
 
-In some cases, you may want to retry at a later time (if for example a third-party provider is down). To do so, your plugin can throw a `NotificationPluginApiRetryException` to include its own retry schedule. When the `NotificationPluginApiRetryException` is caught by Kill Bill, the system computes the next retry date based on the schedule specified in the exception and the number of times that specific event has been retried. In terms or responsabilities:
+== How to develop your custom notification plugin
 
-* Plugin is in charge of deciding whether an `NotificationPluginApiRetryException` should be thrown and attach the associated retry schedule to it.
-* Kill Bill manages the retry logic and also keeps count of # exiting retries versus retry schedule.
+We provide an interface called https://github.com/killbill/killbill-platform/blob/master/osgi-bundles/libs/killbill/src/main/java/org/killbill/billing/osgi/libs/killbill/OSGIKillbillEventDispatcher.java[OSGIKillbillEventDispatcher.OSGIKillbillEventHandler]. In order to create your own notification plugin, you need to write code that implements this interface and implement the `public void handleKillbillEvent(final ExtBusEvent killbillEvent)` method. 
+
+We also provide a simple https://github.com/killbill/killbill-hello-world-java-plugin[Hello World Plugin] that can be used as the starting point to develop your custom notification plugin. 
+You can refer to our https://docs.killbill.io/latest/plugin_development.html#_java_plugins[plugin development] document in order to understand how to set up the https://github.com/killbill/killbill-hello-world-java-plugin[Hello World Plugin] plugin code.
+
+The https://github.com/killbill/killbill-hello-world-java-plugin[Hello World Plugin] has a class called https://github.com/killbill/killbill-hello-world-java-plugin/blob/master/src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldListener.java[HelloWorldListener.java]. This implements the `OSGIKillbillEventDispatcher.OSGIKillbillEventHandler` interface . You can create a class similar to this and write your plugin code within the `handleKillbillEvent` method.
+
+Kill Bill provides a class called https://github.com/killbill/killbill-platform/blob/master/osgi-bundles/libs/killbill/src/main/java/org/killbill/billing/osgi/libs/killbill/OSGIKillbillAPI.java[OSGIKillBillAPI.java]. This class exposes all of Kill Bill's internal APIs. An object of this class will automatically be available in your plugin. For example, the `HelloWorldListener` declares a field corresponding to `osgiKillbillAPI` at Line 36 which is intialized via the constructor.
```

Comment:
Instead of specifying the line number, you could also specify a permanent link here instead. On GitHub, you can click on a line number to highlight it. If you press `y` again, this will generate a permanent link with the line number as well. That way, if we shuffle the code in that class, the reader isn't lost :-)

Reply:
Fixed